### DISCUSSION TITLE
TypeVar "Ts" appears only once in generic function signature

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [microsoft/pyright#3953] [Q&A]
Can someone please explain to me what this error means? 

For reference, I am attempting to type hint `*args` as suggested in [PEP 646](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) and am getting this error [on the pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIALm3I2MC_1WPzWrDMBCE73qKvUkqorQ9GnzoOzi9BCPUZOWKRCuhH4gxfvdKxTl0brv7DbNjU_Cgta2lJtQanI8hFTBEoZjiAmXGbGfKGh0tGh8FKff9E53WiF8mTTXeUcGJorncGJsyjP9Ogk-ZS8auaMGGIF5MWvJw8OfSkfMxTHmeZzkwaLIhAV0fChoOjgCpekymoOj-A-qKyVERln-mZWuGfYCtEbuCrX3-R8u95zvb6pLxvew4AtfaG0da82deEO8KPl7fFAj-XS-3O4JfIf8E5FL-Av3ze2EvAQAA) at line 7.

```python
from __future__ import annotations

from typing_extensions import TypeVarTuple, Unpack

Ts = TypeVarTuple('Ts')

def foo(*args: Unpack[tuple[Unpack[Ts]]]):  # Error occurs here
    for ndx, arg in enumerate(args):
        print(f'Arg{ndx}: {arg}, {type(arg)}')

if __name__ == '__main__':
    foo(1, 2.0, ('buckle my shoe'))
```

**Comment by hmc-cs-mdrissi** [comment]
We can simplify this. This code,

```python
from typing import TypeVar

T = TypeVar('T')

def foo(x: T):
  print('hi')
```

produces same kind of error. The main value of generic type is when you will use it in multiple places. Typically if it's used as both input type and part of return type. It could also just appear several times in input type. It can also be fine if type variable is under class scope.

But a bare function with a single type variable is rarely useful. If you don't care what type is you should just use object. For your code,

```python
def foo(*args: object): 
    for ndx, arg in enumerate(args):
        print(f'Arg{ndx}: {arg}, {type(arg)}')
```

is recommended way to type it.

**Reply by adam-grant-hendry** [reply]
> The main value of generic type is when you will use it in multiple places...It could also just appear several times in input type.

`PEP 646` does give a few examples where a `TypeVarTuple` is used only once:

```python
def execle(path: str, *args: *Tuple[*Ts, Env]) -> None: ...
```

and it states

> `*args` is the only case where an argument can be annotated as `*Ts` directly

```python
def foo(*args: Tuple[*Ts]): ...

foo((0,), (1,))    # Valid
foo((0,), (1, 2))  # Error
foo((0,), ('1',))  # Error
```

Hence, when attempting to type hint a 3rd-party library (where source can't be rewritten) or in subclassing a builtin container type that also uses `*args`, like the [Python 3.8 docs `LRU` cache example](https://docs.python.org/3.8/library/collections.html#ordereddict-examples-and-recipes)

```python
class LRU(OrderedDict):
    'Limit size, evicting the least recently looked-up key when full'

    def __init__(self, maxsize=128, /, *args, **kwds):
        self.maxsize = maxsize
        super().__init__(*args, **kwds)

    def __getitem__(self, key):
        value = super().__getitem__(key)
        self.move_to_end(key)
        return value

    def __setitem__(self, key, value):
        if key in self:
            self.move_to_end(key)
        super().__setitem__(key, value)
        if len(self) > self.maxsize:
            oldest = next(iter(self))
            del self[oldest]
```

It does seem fine to only have `Ts` appear once. I seems like `pyright` shouldn't emit a warning in the case when `*args` is being typed as this is an acceptable use case demonstrated by the PEP.

> ...is recommended way to type it.

My example was trivial, but I appreciate your help. My ultimate goal was to understand the purpose of the warning. It seems to be a false positive (or at least unnecessary) in the event one is trying to type hint an indeterminant number of heterogenous types.

**Reply by erictraut** [reply]
> It does seem fine to only have Ts appear once.

It's "fine" in the sense that it doesn't do any harm, but it doesn't add any value, and it's generally indicative of a user's misunderstanding of how type variables work. If the use of a TypeVar is unnecessary and provides no value in a particular case, you are probably doing something wrong. As @hmc-cs-mdrissi mentioned above, it is equivalent to use `object` or `Any` in this case.

We had a discussion about a year ago in the typing-sig forum about whether a type checker should emit a warning in this case. IIRC, pytype, pyre and pyright all decided to emit a warning, but mypy decided against it. I was initially opposed to it, but I changed my mind after responding to dozens of questions and invalid bug reports from pyright users.

If you would like to disable this warning, you're welcome to do so. It's controlled by the `reportInvalidTypeVarUse` diagnostic rule.

**Reply by hmc-cs-mdrissi** [reply]
Is there any difference in,

```python
def foo(*args: *tuple[*Ts]):
```

vs

```python
def foo(*args: *Ts):
```

Latter I don't know a case where it makes a difference vs `def foo(*args: object):`. Your original example was this form. On other hand,

```python
def bar(*args: tuple[*Ts]):
````

does define some constraint based on your examples and makes sense to allow by itself. It may make sense to drop rule entirely if distinguishing it being useful vs not is complex. It could also be considered too niche given warning is usually valid.

Even for standalone normal typevars there are special cases where T vs object matters, but they are somewhat esoteric and warning emits on them too and I just ignore it when needed.

**Reply by adam-grant-hendry** [reply]
> If you don't care what type is you should just use object.

"Don't care" is a bit strong, but I do agree that `object` is a fair choice if the method is doing little more than swallowing `*args`. I think I was concerned about losing type information, but the types would probably have to be narrowed later where used, and in the case of setting variables (e.g. `__setitem__()`), generics would be in multiple places in the signature (to your point).

**Reply by adam-grant-hendry** [reply]
> It's "fine" in the sense that it doesn't do any harm, but it doesn't add any value

It is frustrating when [PEPs](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) provide examples of doing things in this manner and the checker issues errors and warnings when it's used, only to find later that it's "not best practice". This makes learning proper usage really difficult.

> it's generally indicative of a user's misunderstanding of how type variables work. If the use of a TypeVar is unnecessary and provides no value in a particular case, you are probably doing something wrong.

I understand what you are saying is not directed specifically at me, but that you mean the warning is there to govern the majority case, which is that most often this is an indication that the user is doing things "incorrectly".

However, there really are 2 camps of those doing typing work:

1. Those that have to type hint legacy libraries, specifically those ported from C that are not very Pythonic to begin with, where the end user is stuck and there's not an opportunity to do it in a nice/clean way.

2. Those who are writing with python from the beginning, either using clean/pythonic libraries or writing from scratch from the beginning, where the end user has the greatest leverage to do things "properly"

I utterly begrudgingly belong to the first camp and desperately wish I could do things "better", but I often have to employ ugly ways to get things working.

I am also struggling to find a solid reference for typing best practices with respect to Python's typing system, but I'm not able to find anything. It's maddening to read, re-read, and read again every bit of typing documentation you can find, scour all github issues, and still come up short on a solid understanding of these typing concepts.

**Reply by adam-grant-hendry** [reply]
My comments are simply an expression of my personal frustration attempting to learn and understand typing concepts and have absolutely nothing to do with either you @erictraut , @hmc-cs-mdrissi . You have been wonderful and I am indebted to you for your help. I just wish I knew how to learn these things better because I feel I can't find any other learning options. I just wanted you to know I'm thankful for all your help.

**Reply by hmc-cs-mdrissi** [reply]
The best issues on this topic, typing best practices, is probably [here](https://github.com/python/cpython/issues/91533) and [here](https://github.com/python/typing/issues/845). There's an idea I agree advanced typing docs could use improvement and often I've learned these things by a mix of reading too many issues/local testing and debugging. Progress moves at pace of open source, tends to be when maintainers have time but we'd definitely appreciate discussion/prs for improving typing documentation.

**Reply by gh-andre** [reply]
There's a practical side of being able to use type aliases to simplify function signatures. `MongoClient`  is a good example, which requires `MongoClient[Dict[str, Any]]` used when it is passed into functions to avoid IntelliSense errors. 

For example, aliasing `XY[int int]` as `XYNUM` makes it more clear what `xynum_fn` is supposed to work on and also keeps it more concise.

```python
from typing import TypeVar, Generic

XT = TypeVar("XT")
YT = TypeVar("YT")

class XY(Generic[XT, YT]):
    def __init__(self, x: XT, y: YT):
        self.x: XT = x
        self.y: YT = y

XYNUM = TypeVar("XYNUM", bound=XY[int, int])

def xynum_fn(xynum: XYNUM) -> int:              # reports "appears only once"
    return xynum.x * xynum.y

num: int = xynum_fn(XY[int, int](123, 456))     # XY is typically allocated elsewhere

print(num)
```
It's essentially like a type alias in C++ and having it in parameters alone (i.e. "appears once") is useful for type checking.

**Reply by erictraut** [reply]
Yes, type aliases have value, but what you're doing here is defining a bound TypeVar, not a type alias. You can define a type alias simply as `XYNUM = XY[int, int]`.

**Reply by gh-andre** [reply]
Thanks for responding. 

With `TypeVar` I also get additional control for constraints and covariance, though.  Maybe I'm confusing the intent, but it seems to me that `TypeVar` is intended for more granular way to specify parameter types in generic functions.

**Reply by erictraut** [reply]
Variance applies only when a `TypeVar` is used in a class, not in a generic function or type alias. If a function-scoped `TypeVar` appears only once in the function's signature, you're almost definitely using it inappropriately.

**Reply by gh-andre** [reply]
Let me focus on the reported error and point out that a helpful error message would specify what's actually wrong, so it can be addressed, but instead it just says, rather cryptically, that `TypeVar` appears only once, concluding that the author "is almost definitely using it inappropriately".

**Reply by erictraut** [reply]
It's hard to say what's wrong in this case other than a lack of understanding of how TypeVars are meant to be used.

**Reply by sassanh** [reply]
Imagine this:

```python
from typing import TypeVar

T = TypeVar('T')
U = TypeVar('U')

def foo(x: T, y: U):
  a: T = ... # Somehow initialized as T
  b: U = ... # Somehow initialized as U
  c: U = ... # Somehow initialized as U
  ...

  c = a # If I mistakenly assign a to c, it raises an error, because instances of T are not necessarily assignable to instances of U
  print('hi')
```

If I used `Any` or `Unknown` in place of `T` and `U` for typing `x` and `y`, I wouldn't see any typing error. So I think there is value in using a TypeVar even only once in function signature: to distinguish different types in function body.

**Reply by gh-andre** [reply]
Having two different types is another good example of using types for achieving useful side effects within some function, rather than returning a result of applying one of those types, as the warning suggests (well, sort of).

Similarly, calling type's methods would be one more, like this:
```python
from typing import TypeVar
from abc import abstractmethod

class A:
    @abstractmethod
    @staticmethod
    def param_x() -> int: ...

T = TypeVar("T", bound=A)

def fn(p: T) -> None:
    print(p.param_x)

# somewhere else

class B(A):
    @staticmethod
    def param_x() -> int:
        return 123

fn(B())
```

**Reply by Azureblade3808** [reply]
Why not just write
```python
def fn(p: A) -> None:
    print(p.param_x)
```
?

**Reply by gh-andre** [reply]
This is just an example. `T` may be defined as a union of types or various type constraints, etc.
```python
T = TypeVar("T", bound=SupportsAbs[float])

def fn(p: T) -> None:
    print(abs(p))
```

**Reply by Azureblade3808** [reply]
```python
T: TypeAlias = SupportsAbs[float]

def fn(p: T) -> None:
    print(abs(p))
```
Am I still missing something here?

**Reply by erictraut** [reply]
@Azureblade3808, no, you're not missing something.

@gh-andre, all of your examples above demonstrate cases where a type variable is unnecessary. Any type specified in a type annotation already implies all subtypes of that type. For example, the type annotation `A` implies "an instance of `A` or any subtype thereof". If `B` is a subclass of `A`, then an instance of `B` is compatible with the type annotation `A`. There's no need to use a bound TypeVar to indicate as much.

**Reply by gh-andre** [reply]
@erictraut 
> all of your examples above demonstrate cases where a type variable is unnecessary.

On the contrary. `TypeVar` indicates that the type is intended to be the actual type and that this code is not intended to use inheritance, but rather a template mechanism to allow parameters to be constrained in specific ways. This pattern is common in other languages, like C++, and it's important to distinguish these cases.

This warning fails to deliver the reason of why it is being reported. If you take any other warning, one can describe the negative outcome of not adhering to what is being highlighted, such as a bad access via `None`, etc. This one just describes the code - the type being used once in the signature. It's no different than reporting code formatting inconsistencies as warnings.

**Reply by erictraut** [reply]
> On the contrary.

No, the examples you've provided above are all cases where the use of a type variable is completely unnecessary. It provides zero value. You can simply replace the use of the TypeVar with its upper bound constraint, and the effects will be the same.

> This warning fails to deliver the reason of why it is being reported.

This warning is effectively saying "you are using a feature of the type system in a way that indicates you don't understand something; the results may not be what you intend". It is meant to prompt users to educate themselves so they can avoid mistakenly (or unnecessarily) using a feature of the type system as you're doing in your examples above.

It's difficult within a diagnostic message to provide a full explanation of how TypeVars should be used. If you can think of a better way to word the warning's text, I'm open to suggestions.

**Reply by Azureblade3808** [reply]
```plain
TypeVar "Ts" should be replaced with some type since it appears only once in generic function signature.
```
How about this?

or
```plain
TypeVar "Ts" is meaningless since it appears only once in generic function signature.
```

**Reply by erictraut** [reply]
OK, I've augmented the diagnostic to include a suggested fix.

![image](https://github.com/microsoft/pyright/assets/7040122/56b4e88b-c531-4534-b9d5-30d165c782f2)

**Reply by sassanh** [reply]
My comment is buried behind https://github.com/microsoft/pyright/discussions/3953#discussioncomment-7605768
Is it also possible to distinguish between different types no necessarily with different bases without TypeVars?

**Reply by Azureblade3808** [reply]
> ```python
> from typing import TypeVar
> 
> T = TypeVar('T')
> U = TypeVar('U')
> 
> def foo(x: T, y: U):
>   a: T = ... # Somehow initialized as T
>   b: U = ... # Somehow initialized as U
>   c: U = ... # Somehow initialized as U
>   ...
> 
>   c = a # If I mistakenly assign a to c, it raises an error, because instances of T are not necessarily assignable to instances of U
>   print('hi')
> ```

> Is it also possible to distinguish between different types no necessarily with different bases without TypeVars?

I'm afraid that it is impossible.

But in the other hand, why would you want to forbid `c = a`? How are you going to use `c`, so that it will cause you trouble if type of `c` is `T` instead of `U`?

**Reply by sassanh** [reply]
> But in the other hand, why would you want to forbid `c = a`? How are you going to use `c`, so that it will cause you trouble if type of `c` is `T` instead of `U`?

It doesn't cause me trouble if the type is changed, it causes me trouble if I'm mistaken about the role of `a` while writing that line and the static typing system is expected to remind me I'm doing something wrong, I may decide to explicitly cast a to U and let the assignment be, but I should be informed by my editor that I'm doing implicit type casting and something may be wrong, just how it is the case in other languages with static typing.

Take a look at this code:

```python
from typing import Callable, TypeVar


State = TypeVar("State")
CacheKey = TypeVar("CacheKey")
Value = TypeVar("Value")


def generate_value_only_if_cache_key_changed(
    state: State,
    cache_key_generator: Callable[[State], CacheKey],
    value_generator: Callable[[State], Value],
):
    last_cache_key = None
    last_value = None

    def check_results_and_generate_value_if_needed():
        nonlocal last_cache_key, last_value
        cache_key = cache_key_generator(state)
        if cache_key != last_cache_key:
            try:
                last_cache_key = cache_key
                value = value_generator(state)
                return value
            finally:
                last_value = cache_key # Bad stuff here
        return last_value

    return check_results_and_generate_value_if_needed
```

I write this code solely to demonstrate the use of single usage of TypeVar in function signature, it is not tested, so let's just focus on the issue.

Pyright currently generates this:

```
/private/tmp/g/a.py
  /private/tmp/g/a.py:28:30 - error: Expression of type "CacheKey@generate_value_only_if_cache_key_changed" cannot be assigned to declared type "Value@generate_value_only_if_cache_key_changed | None
"
    Type "CacheKey@generate_value_only_if_cache_key_changed" cannot be assigned to type "Value@generate_value_only_if_cache_key_changed | None"
      Type "CacheKey@generate_value_only_if_cache_key_changed" cannot be assigned to type "Value@generate_value_only_if_cache_key_changed"
      "object*" is incompatible with "None" (reportGeneralTypeIssues)
  /private/tmp/g/a.py:13:44 - warning: TypeVar "CacheKey" appears only once in generic function signature (reportInvalidTypeVarUse)
  /private/tmp/g/a.py:14:40 - warning: TypeVar "Value" appears only once in generic function signature (reportInvalidTypeVarUse)
1 error, 2 warnings, 0 informations 
```

I think the error here is a valuable error, it helps the user to find a mistake in their code before even running it to test it. It is one of the main objectives of static typing to find such problems in the code. So I think the two warnings generated here are irrelevant.

**Reply by Azureblade3808** [reply]
I see what you are trying to accomplish. While isolated type-var usually indicates programming errors, I think it may be good in your case. It's just not what type-vars are designed for. I'd suggest just using a `# type: ignore` to suppress the warning.

Back on the example you gave, it is just not convincing enough. If you didn't omit the return type of `generate_value_only_if_cache_key_changed`, its full signature would be something like
```python
def generate_value_only_if_cache_key_changed(
    state: State,
    cache_key_generator: Callable[[State], object],
    value_generator: Callable[[State], Value],
) -> Callable[[], Value]:
    ...
```
You won't get a complain from Pyright or mistakenly assign a cache key to a value with it. Maybe you could give a better example for other readers.

**Reply by sassanh** [reply]
I think this rule in its current form is not correct. A single counterexample is enough to show a general rule is not "general". If we change the above function to print the value or store it in a database instead of returning it then the return value of the wrapper function becomes `Callable[[], None]` instead of `Callable[[], Value]`.
So considering this rule is not generally applicable to all cases, I think either it should be evolved to take care of the body of the function to make sure the `TypeVar` is actually not needed or it should be disabled by default.
I would like to know the opinion of a maintainer too. @erictraut

**Reply by Azureblade3808** [reply]
By the way, if you are mainly concerned in variable roles, `typing.NewType` may be what you need.

**Reply by sassanh** [reply]
Instead of guessing I would like to know the complete rational behind this statement I'm hearing here implicitly: "if a TypeVar is used only once in a function signature, something is wrong, you don't need a TypeVar used only once."
It is the conclusion of a reasoning, I'm interested to learn about the reasoning, is it documented somewhere?

**Reply by gh-andre** [reply]
> It is the conclusion of a reasoning, I'm interested to learn about the reasoning, is it documented somewhere?

Second point in this comment
https://github.com/microsoft/pyright/discussions/3953#discussioncomment-7625615

**Reply by sassanh** [reply]
> Second point in this comment
> [#3953 (reply in thread)](https://github.com/microsoft/pyright/discussions/3953#discussioncomment-7625615)

If you mean this:

> This warning is effectively saying "you are using a feature of the type system in a way that indicates you don't understand something; the results may not be what you intend". It is meant to prompt users to educate themselves so they can avoid mistakenly (or unnecessarily) using a feature of the type system as you're doing in your examples above.

I don't see the rational being explained here nor it has any reference to an external document.

**Reply by erictraut** [reply]
When a feature of the language causes frequent confusion and results in incorrect usage, it's good for tools to try to help steer people in the right direction. That's the purpose of this rule.

I'll reiterate that if you don't like the `reportInvalidTypeVarUse` rule, you can disable it. It's enabled by default (as a warning) for a good reason. Python developers who are new to generics often misuse TypeVars or use them in cases where they are not the right tool for the job. This is very common. I also see it with developers who _think_ they understand TypeVars but don't. 

@sassanh, you've provided two examples. The first one seems very contrived, and I don't know what you're trying to do in it. The second one doesn't appear to demonstrate the point you're trying to make. Do you have any better examples?

**Reply by sassanh** [reply]
Thanks for explaining it! I also found this and I put its link here as it may help future readers get some context about the issue: https://mail.python.org/archives/list/typing-sig@python.org/thread/NRFNHGXHXPGBR6FP3TIOZZ6VS4XJZX6K/

I understand that one may assume the nature of developing a static type checker is to help people with real world scenarios so you may find my "contrived" example and my other "can't find the real world benefit of this piece of code" example are not helpful here. But we can see the issue from a different point of view too, a purely logical point of view, in which a single counter-example - no matter if it is contrived or if it has any real world uses or not - is enough to question a general rule. And if a rule is not a general rule then a static type checker becomes an opinionated static type checker if it decides to have it turned on by default.

It's ok for a linter to be opinionated (like prettier or many others) but for a static type checker I think it's a bit unusual to report "I think probably you didn't want this" kind of warnings. As a suggestion you may consider having an option to turn off all the opinionated rules altogether as some developers prefer their static type checkers to report factual mistakes than "probably not ok, if you don't know what you are doing you better double check" cases.

I think the problem here is that a more sophisticated, factual and probably harder to implement rule "you are not using any features of a TypeVar here" has been simplified to a "since you are using this TypeVar only once in your function signature, we guess you are not using any features of a TypeVar here".

**Reply by erictraut** [reply]
@sassanh, I understand your perspective. In this particular case, I feel pretty strongly about leaving this warning in place because it's serving an important purpose. As someone who spends many hours every week answering typing-related questions, I'm motivated to find ways for pyright to help guide users away from common pitfalls and toward correct usage of type-related features.

I still haven't seen a non-contrived use for a regular `TypeVar` (one without constraints) or a `TypeVarTuple`. If you can show me one, then we could think about exemptions that would suppress this warning for that case. If your argument is purely theoretical, then I don't think it's worth our time debating this further.

**Reply by gh-andre** [reply]
@sassanh 
In your specific example, you can probably use something like this, which might be useful in a context of a cache class. It won't emit this warning either, but just to be clear, I'm not suggesting to restructure code just to suppress a warning - a wrapper class may just be useful on its own.

```python
from typing import TypeVar, Generic, Callable

State = TypeVar("State")
CacheKey = TypeVar("CacheKey")
Value = TypeVar("Value")

class StatefulCache(Generic[Value, CacheKey]):
    def generate_value_only_if_cache_key_changed(self, 
        state: State,
        cache_key_generator: Callable[[State], CacheKey],
        value_generator: Callable[[State], Value],
    ) -> None:
        ...
```

**Reply by sassanh** [reply]
@erictraut I have a real world example [here](https://github.com/sassanh/python-redux/blob/main/redux/__init__.py#L115).
Pyright is reporting a warning for this line for `ComparatorOutput`, I don't want to type it `Any` or `object` because I don't want to mistakenly assign a selector output to a variable with type `ComparatorOutput` and I want all such assignments be explicit and in controlled branches of code, one such explicit cast is [here](https://github.com/sassanh/python-redux/blob/main/redux/__init__.py#L135).

I think even without a real world example it is expected that a TypeVar may be used only once in the signature, but may be used in the function body to distinguish its variables from other variables typed as another TypeVar or non-variable types. This is to help the developer avoid mistakenly assigning variables of different types to each other.

I appreciate your time, I'm already happy disabling `reportInvalidTypeVarUse` in my codebase as you suggested. The only reason I'm continuing this discussion is in the hope that I'm bringing some value to Pyright. Please feel free to leave it as is if you think this is not helpful.

**Reply by sassanh** [reply]
@gh-andre Thanks! Yeah this is a good workaround, there are other workarounds and hacks that can silence the warning too, but silencing the warning is not my primary goal :) I'm trying to find out the rational behind this warning (which now I know thanks to Eric's explanation) and try to improve the situation upstream if possible.

**Reply by erictraut** [reply]
@sassanh, OK I agree that's a legit use. Thanks for the example. I've found a way to adjust the heuristic so it allows this case while still flagging the common misuses with a warning. I think this is a reasonable compromise.

**Reply by Azureblade3808** [reply]
@erictraut Your new algorithm may be too permissive IMHO.

When I changed the related test sample from
```python
def callable(f: Callable[P, R]):
    def inner(*args: P.args, **kwargs: P.kwargs) -> list[R]:
        return [f(*args, **kwargs)]

    return inner
```
into
```python
def callable(f: Callable[P, R]):
    def inner(*args: P.args, **kwargs: P.kwargs):
        pass

    return inner
```
There is still no warning even when `R` is totally meaningless here.

While I still haven't made sure how to implement it, I think an alternative way could be scanning if a type-var is used only once through the function in both its interface and implementation. This should have a result like:
```python
# Good. `T` is used twice.
def f0[T](x: T) -> None:
    y: T = x  # pyright: ignore[reportUnusedVariable]

# Bad. `T` is used only once, when `y` doesn't have an explicit type.
def f1[T](x: T) -> None:
    y = x  # pyright: ignore[reportUnusedVariable]

# Good. Both `P` and `R` are used more than once.
def callable[**P, R](f: Callable[P, R]):
    def inner(*args: P.args, **kwargs: P.kwargs) -> list[R]:
        return [f(*args, **kwargs)]

    return inner

# Bad. `R` is used only once.
def callable[**P, R](f: Callable[P, R]):
    def inner(*args: P.args, **kwargs: P.kwargs):
        pass

    return inner

# Also bad. `R` is used only once, when neither `callable` nor `inner` has an explicit return type.
def callable[**P, R](f: Callable[P, R]):
    def inner(*args: P.args, **kwargs: P.kwargs):
        return [f(*args, **kwargs)]

    return inner
```

**Reply by erictraut** [reply]
Yes, the new algorithm is more permissive. I'm trying to strike a good balance. The important thing is to cover the most common misuses hit by developers who are just starting to use TypeVars. Someone who is using a `ParamSpec` is far beyond the beginning phase.

**Reply by sassanh** [reply]
Thanks for the quick action! I guess I found another case when single appearance of a `TypeVar` is legit:

While this already works without `TypeVar` and inheritance is enough (note that `A` is the return value type of the callable `test` takes as input):
```python
from typing import Callable
class A:
    ...
class B(A):
    ...
def test(fn: Callable[[], A]) -> None:
    ...
def f() -> B:
    return B()
test(f)
```

This doesn't work with inheritance (note than `A` is in the arguments of the callable `test` takes as input)

```python
from typing import Callable
class A:
    ...
class B(A):
    ...
def test(fn: Callable[[A], None]) -> None:
    ...
def f(value: B) -> None:
    ...
test(f)
```

I think I understand the rationale behind this behavior: `test` shouldn't be able to call `fn` with an `A` instance as it needs a `B` instance and `B` may have more features than `A`, but it's ok if `test` calls `fn` and get a `B` instance while it is expecting an `A` instance since `B` has all the features of `A`.

So I think for the second scenario the signature of `test` should be defined with a `TypeVar` bound to `A`. I wanted to share this with you.
Let me know if a real world example is needed.

**Reply by erictraut** [reply]
Input parameter types are contravariant. How are you proposing to implement `test` in your second example above? How would it know what type the callback function can accept if you used a type variable in the `Callable` annotation?

```python
def test[T: A](fn: Callable[[T], None]) -> None:
    fn(A()) # Not type safe!
    fn(B()) # Not type safe!
```

The only way this would make sense is if you passed a value of type `T` to `test` as well.

```python
def test[T: A](fn: Callable[[T], None], arg: T) -> None:
    fn(arg) # Type safe!

test(f, B())
```

**Comment by gh-andre** [comment]
> This warning is effectively saying "you are using a feature of the type system in a way that indicates you don't understand something; the results may not be what you intend". 

You might as well use this phrase as the error message - it's just as informative, but far more honest. Certainly works better than suggesting to use `object` or `Any`.

**Reply by Azureblade3808** [reply]
We are not in the position to judge the user.

Such warnings can be caused by misunderstanding of how `TypeVar` works, or typos, or unfinished code, or other reasons. It could be very insulting to say "you don't understand something" or similar words.

**Comment by sytranvn** [comment]
I want to use TypeVar to declare a generic type that would be extended in subclass like a config class with extra fields.
```py
# foo.py
class Foo:
    a: int
    b: int
```

```py
# bar.py
from abc import ABC, abstractmethod
from typing import TypeVar

from foo import Foo

FooBase = TypeVar("FooBase", bound=Foo)


class Bar(ABC):
    @abstractmethod
    @classmethod
    def init_foo(cls, foo: FooBase):
        #                  ~~~~~~~
        # TypeVar "FooBase" appears only once in generic function signature
        #  Use "Foo" instead [reportInvalidTypeVarUse]
        pass

```

```py
# subbar.py
from bar import Bar
from foo import Foo


class SubFoo(Foo):
    c: str


class SubBar(Bar):
    @classmethod
    def init_foo(cls, foo: SubFoo):
        return super().init_foo(foo)
```
But if I replace `FooBase` with `Foo` in `Bar.init_foo`, I get error message in override method of subclass.
```
Method "init_foo" overrides class "Bar" in an incompatible manner
     Parameter 2 type mismatch: base parameter is type "Foo", override parameter is type "SubFoo"
       "Foo" is not assignable to "SubFoo" [reportIncompatibleMethodOverride]
```

**Reply by erictraut** [reply]
I don't have the full context for the problem you're trying to solve, but it looks from your code sample that you want a class-scoped type variable rather than a method-scoped type variable.

```python
class Bar[FooBase: Foo](ABC):
    @abstractmethod
    @classmethod
    def init_foo(cls, foo: FooBase):
        pass

class SubBar(Bar[SubFoo]):
    @classmethod
    def init_foo(cls, foo: SubFoo):
        return super().init_foo(foo)
```

**Reply by sytranvn** [reply]
Yes, this is what I want to achieve. But this syntax only works with python>=3.12 but unfortunately our codebase is still python3.10.

**Reply by erictraut** [reply]
You can use the old syntax to achieve the same thing. The important point is that it should be a class-scoped TypeVar, not a method-scoped TypeVar.

**Reply by sytranvn** [reply]
```py
from abc import abstractmethod
from typing import Generic, TypeVar


class Foo:
    a: int
    b: int


FooBase = TypeVar("FooBase", bound=Foo)


class BarBase:
    pass


class Bar(BarBase, Generic[FooBase]):
    @abstractmethod
    @classmethod
    def init_foo(cls, foo: FooBase):
        pass


class SubFoo(Foo):
    c: str


class SubBar(Bar):
    @classmethod
    def init_foo(cls, foo: SubFoo):
        return super().init_foo(foo)
```

I put the solution to my problem here in case anyone ever needs it.
